### DISCUSSION TITLE
Using libfido2 on Windows without Windows Hello

### DISCUSSION THREAD

**Discussion by apantina** [Yubico/libfido2#699] [Q&A] [answered]
Hi folks!

I am trying to use your library for WebAuthn flows outside of browsers. Also I am limited to not using Windows Hello and would prefer to talk to the authenticator directly.

The flow is fairly simple - get a challenge from a relying party, call `fido2-token` to find a device and its handle, and then use that handle to call something like `cat challenge_data.txt | fido2-assert -G <device>` with the above device handle and challenge data. This flow works great on Mac & Linux by the way, so thanks for the support and work done on this library. :)

However on Windows I get no output if I try running `fido2-token`.
Debug info:
```
PS C:\Users\ardi> fido2-token.exe -L -d
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: HidD_GetPreparsedData
is_fido: unsupported report len

```
Interestingly enough, if I run the above command in an elevated PowerShell (admin rights) I am able to get a device handle. Also I can then use the same handle to call `fido2-assert` and generate an assertion successfully. 
In an elevated shell:
```
PS C:\Users\ardi> fido2-token.exe -L -d
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: unsupported report len
is_fido: HidD_GetPreparsedData
is_fido: unsupported report len
\\?\hid#vid_1050&pid_0407&mi_01#7&21a877f4&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}: vendor=0x1050, product=0x0407 (Yubico YubiKey OTP+FIDO+CCID)
```
I am using the latest release (1.13.0) downloaded from https://developers.yubico.com/libfido2/Releases/ on Windows 10 Enterprise 22H2. 
Am I doing something entirely wrong here? Thanks in advance.

**Comment by martelletto** [comment, marked as answer]
Hi @apantina,

The only way to interface with a FIDO device from an unprivileged context in recent versions of Windows 10 and in Windows 11 is through Windows Hello. This is a OS-level requirement.

It's a bit strange that you aren't seeing Windows Hello in the output of `fido2-token -L`. Did you explicitly disable `USE_WINHELLO` when building?

Thank you,

-p.

**Comment by apantina** [comment]
Hi @martelletto, thanks for the quick response!
Unfortunately Windows Hello is disabled on work machines by my employer for security reasons (mostly related to PINs). I used default build flags so `USE_WINHELLO` was turned on.

**Comment by martelletto** [comment]
That's quite interesting. Is there an alternative mechanism for unprivileged applications (e.g. Chrome) to communicate with FIDO keys? If so, we could look into supporting it in libfido2.

**Comment by apantina** [comment]
Browsers such as Chrome and Edge work fine with FIDO2 keys via WebAuthn even with Windows Hello disabled on their respective machines.
While I am not very familiar with the underlying mechanisms, this is what pops up when I use that authentication flow via Chrome. 
![image](https://github.com/Yubico/libfido2/assets/110903292/1beeb033-15fc-46c7-bc68-6185cce49cab)
I can confirm that this flow uses FIDO2 (although I'm not showing the entire page and I obfuscated some parts of the above screenshot :)). 

Looking into Chromium code [here](https://chromium.googlesource.com/chromium/src/+/HEAD/device/fido/win), it seems like they use their WebAuthN [header file](https://github.com/microsoft/webauthn/blob/master/webauthn.h) and [API](https://learn.microsoft.com/en-us/windows/win32/webauthn/-webauthn-portal
), but I don't see any direct references to Windows Hello.

**Comment by martelletto** [comment]
Oh, I see. Thanks. That dialogue comes from Microsoft's webauthn.dll, which is usually (but mistakenly) referred to as Windows Hello. We support webauthn.dll in libfido2, so it's strange that `fido2-token -L` did not list it as an option. It also doesn't appear to have failed, given the absence of error messages in that path. It's unclear what happened. From the debug output, it looks as if libfido2 was built without `USE_WINHELLO`.

**Comment by apantina** [comment]
Thanks for the clarification! For the record, I used pre-built binaries from the [Releases](https://developers.yubico.com/libfido2/Releases/) page so I assume they had the default `USE_WINHELLO` value (`true`). I can have a go at building them myself as well just in case.

**Comment by martelletto** [comment]
An additional data point: @LDVG had a look and confirmed that the latest release image for Windows (1.13.0) was built with `USE_WINHELLO`.

**Comment by apantina** [comment]
Going back to your previous comment: 
> We support webauthn.dll in libfido2, so it's strange that fido2-token -L did not list it as an option.

Not sure if I follow - what should have been listed here exactly? A device handle or just `"windows://hello"`? I believe that the winhello handle wasn't listed simply because it's **disabled on the machine**, regardless of the fact that the binary was built with `USE_WINHELLO`. 

Should `libfido2` be able to communicate via `webauthn.dll` with FIDO keys in an unprivileged context then? Or is that only possible for Windows Hello? 
Thanks!

**Reply by LDVG** [reply]
The expectation would be that `windows://hello` is listed when running `fido2-token -L`. If we fail to load the backend, an error message should have been logged when using debug output.  Is your pasted output from `fido2-token -L -d` truncated?

Using `windows://hello` as a device handle to e.g. `fido2-assert -G` should bring up the same pop-up that Chrome brings up, in an unprivileged context. The terms Windows Hello and `webauthn.dll` are unfortunately used interchangeably here.

**Reply by apantina** [reply]
> Is your pasted output from fido2-token -L -d truncated?

Nope :) 

> Using windows://hello as a device handle to e.g. fido2-assert -G should bring up the same pop-up that Chrome brings up, in an unprivileged context

I remember trying this, but it didn't work. Don't recall the error message however. I'll reproduce this with debug output once I'm back at my Windows laptop and follow up in about a week's time as I'm currently on PTO :)

**Reply by LDVG** [reply]
> Nope :)

I've looked closer at this now. During a call to `fido2-token -L -d`, version 1.13.0 logs a debug message for each manifest function it runs while enumerating devices:

```console
$ .\fido2-token -L -d
copy_info: path=\\?\hid<truncated>
copy_info: is_fido
run_manifest: found 0 hid devices
webauthn_load: api version 2
run_manifest: found 1 winhello device
windows://hello: vendor=0x045e, product=0x0001 (Microsoft Corporation Windows Hello)
```

Looking at the implementation, a [message should be logged for both successful and failure to run a specific manifest function](https://github.com/Yubico/libfido2/blob/1.13.0/src/dev.c#L243-L278).  Since these log messages are missing in your output, it would seem like another version of libfido2 is picked up on your system. As an additional data point, the "unsupported report len" looks to have been removed in version 1.5.0 (commit 997caf90eb9988ab8b0c9823e7ca9720c4c33e45). 

Do you have another version of libfido2 installed on your system? What is the output of `fido2-token -V`?

**Reply by apantina** [reply]
Hi @LDVG, thanks for getting back to me and apologies for the late reply (wasn't near my laptop). 

So the problem was on my side! I initially installed libfido2 via `scoop`, as instructed elsewhere (https://github.com/keys-pub/go-libfido2 to be specific) which was outdated (`1.4.0`). I mistakenly thought that I was using the latest version from the releases page.

When I run `fido2-token.exe -L -d` on my laptop with the *actual* latest version, I get the following output:
```
$ ./fido2-token.exe -L -d
copy_info: path=\\?\hid#intc816#3&36043c54&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd
copy_info: hid_ok
copy_info: path=\\?\hid#elan0676&col01#5&15198777&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
copy_info: path=\\?\hid#elan0676&col02#5&15198777&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
copy_info: path=\\?\hid#melf0410&col01#5&1dfd0999&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
copy_info: path=\\?\hid#elan0676&col03#5&15198777&0&0002#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
copy_info: path=\\?\hid#melf0410&col02#5&1dfd0999&0&0001#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
copy_info: path=\\?\hid#elan0676&col04#5&15198777&0&0003#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
copy_info: path=\\?\hid#vid_8087&pid_0ac2#6&1f3d8128&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
copy_info: hid_ok
run_manifest: found 0 hid devices
webauthn_load: api version 4
run_manifest: found 1 winhello device
windows://hello: vendor=0x045e, product=0x0001 (Microsoft Corporation Windows Hello)
```

I get identical output whether I run the command as a regular user or as an administrator.

On a slightly unrelated note, if I try creating an assertion via `windows://hello`, I get the following output:
```
$ ./fido2-assert.exe -G -i assert windows://hello -d
fido_dev_get_assert: FIDO_ERR_INTERNAL

```
I can verify that the contents of the `assert` file are fine because I tested them on Mac & Linux devices with no issues. Again, this is likely due to WinHello being disabled on the laptop.

**Reply by LDVG** [reply]
Hi,

> So the problem was on my side! I initially installed libfido2 via scoop, as instructed elsewhere (https://github.com/keys-pub/go-libfido2 to be specific) which was outdated (1.4.0). I mistakenly thought that I was using the latest version from the releases page.

Thanks for checking!

> On a slightly unrelated note, if I try creating an assertion via windows://hello, I get the following output:
>
> ```
> $ ./fido2-assert.exe -G -i assert windows://hello -d
> fido_dev_get_assert: FIDO_ERR_INTERNAL
> ``` 
>
> I can verify that the contents of the assert file are fine because I tested them on Mac & Linux devices with no issues. Again, this  is likely due to WinHello being disabled on the laptop.

Would you mind trying that command again, but moving the `-d` flag to before the `windows://hello` argument (or setting `FIDO_DEBUG=1` in your environment)? 

I believe there is another issue:  our tools predate our `windows://hello` driver and are designed to accept the client data hash, but `webauthn.dll` only accepts the raw client data as input.  We'll have to provide an option in our tools to handle this.

**Reply by apantina** [reply]
Hi @LDVG, here's the output (on 1.13.0): 
```
 ./fido2-assert.exe -G -d -i assert windows://hello
client data hash:
  44 f2 17 df 02 a2 92 94 ca 4e 85 3e be 9a b7 e0
  f9 42 07 9b ad 08 69 75 47 60 42 6f f9 63 45 0f
relying party id: internalfb.com
credential id:
  48 ed c7 13 00 47 1f db c8 ec 3e da 12 82 ab 01
  83 4b 27 df 69 f5 23 18 1c 87 87 2f c1 1c ae 55
  d9 7a 6f e6 14 f4 a3 15 21 15 22 c5 d3 7a d3 5a
  fa 5b c8 91 b5 e8 7e 00 d2 71 68 ca bf 00 0c 4b
up=omit
uv=omit
pin=omit
webauthn_load: api version 4
pack_cd: NULL
translate_fido_assert: pack_cd
fido_winhello_get_assert: translate_fido_assert
fido_dev_get_assert: FIDO_ERR_INTERNAL
```

**Reply by LDVG** [reply]
Thank you. Yes, that'd be the same issue I was referring to.

I just opened a draft PR (#707) where we add a new `-w` option to `fido2-cred -M` and `fido2-assert -G`. This option makes these operations interpret the first line of the input data as base64 blob of the raw client data rather than the client data _hash_  That should work under the `windows://hello` backend; but you'll obviously need to modify your input parameters accordingly. If you don't mind testing that, we'd appreciate feedback. :slightly_smiling_face:

**Reply by apantina** [reply]
Thanks for getting back to me so quickly and for working on this! I checked out your PR and tried building `libfido2` but ran into the following error: 
```
PS C:\Users\ardi> .\libfido2\windows\build.ps1
...
CMake Error at CMakeLists.txt:8 (project):
  Generator

    NMake Makefiles

  does not support platform specification, but platform

    x64

  was specified.


-- Configuring incomplete, errors occurred!
Failed to build LibreSSL
At C:\Users\ardi\libfido2\windows\build.ps1:155 char:2
+     throw "Failed to build LibreSSL"
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Failed to build LibreSSL:String) [], RuntimeException
    + FullyQualifiedErrorId : Failed to build LibreSSL
```

It seems like the wrong CMake generator is being specified, one that doesn't support specifying architectures via `-A`. I dug around a bit but couldn't resolve. Apologies if this is not the appropriate thread for this.

When I run `cmake --help` it says that another (seemingly valid, since it supports `-A`) generator is set:
```
PS C:\Users\ardi> cmake --help
...
Generators

The following generators are available on this platform (* marks default):
* Visual Studio 17 2022        = Generates Visual Studio 2022 project files.
                                 Use -A option to specify architecture.
```

More info: 
```
PS C:\Users\ardi> cmake version
CMake Error: The source directory "C:/Users/ardi/version" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
PS C:\Users\ardi> cmake --version
cmake version 3.26.4

CMake suite maintained and supported by Kitware (kitware.com/cmake).
PS C:\Users\ardi> gcc.exe --version
gcc.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0

PS C:\Users\ardi> g++.exe --version
g++.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0
```
Any idea how I can resolve this? If this is out of scope, could you perhaps provide build artifacts for Windows? I couldn't download them from [Github Actions](https://github.com/Yubico/libfido2/actions/runs/5401767794).

**Reply by LDVG** [reply]
Does setting `CMAKE_GENERATOR="Visual Studio 17 2022"` in the environment help?

**Reply by apantina** [reply]
It helped - I managed to get a successful build, note that I also had to install Visual Studio (not just the libraries but the IDE itself). However running the binaries resulted in new issues coming up:
```
ardi@PC2BDVT2 MINGW64 ~/libfido2 (tools-cd-ci)
$ ./build-Release/tools/fido2-token.exe -d
C:/Users/ardi/libfido2/build-Release/tools/fido2-token.exe: error while loading shared libraries: cygz.dll: cannot open shared object file: No such file or directory

ardi@PC2BDVT2 MINGW64 ~/libfido2 (tools-cd-ci)
$ ./build-Release/tools/fido2-token.exe -L -d
C:/Users/ardi/libfido2/build-Release/tools/fido2-token.exe: error while loading shared libraries: cygz.dll: cannot open shared object file: No such file or directory
```
So some DLLs were not linked even though the `build.ps1` script ran successfully. I ran `cygcheck` to find all of the missing dependencies:
```
C:\Users\ardi\libfido2\build-Release\tools\fido2-token.exe
  C:\WINDOWS\system32\KERNEL32.dll
    C:\WINDOWS\system32\ntdll.dll
    C:\WINDOWS\system32\KERNELBASE.dll
cygcheck: track_down: could not find cygfido2-1.dll

cygcheck: track_down: could not find cygcrypto-1.1.dll

cygcheck: track_down: could not find cygwin1.dll

cygcheck: track_down: could not find cygz.dll

```

I added the following directories to $PATH as these contained the DLLs that were missing:
```
C:\Users\ardi\libfido2\cygwin\root\bin
C:\Users\ardi\libfido2\build-Release\regress
```

After adding these to $PATH I just get empty output from all of the tools, error code 127 (checked with `echo $?`). Am I running the wrong artifacts or is something else the issue? Thanks.

**EDIT** - if  I run an artifact from `build` instad of `build-Release` I get the following output for `fido2-token`:
```
ardi@PC2BDVT2 MINGW64 ~/libfido2 (tools-cd-ci)
$ ./build/x64/dynamic/tools/Release/fido2-token.exe -d -L
C:/Users/ardi/libfido2/build/x64/dynamic/tools/Release/fido2-token.exe: error while loading shared libraries: api-ms-win-crt-locale-l1-1-0.dll: cannot open shared object file: No such file or directory
```

**Reply by LDVG** [reply]
Apologies for the extra troubles. As a workaround, assuming you're building with shared libraries turned on, can you run
```
Get-ChildItem -Path "..\..\regress" -Include *.dll -Recurse | Copy-Item
```
from your `libfido2\build\x64\dynamic\tools\Release` directory and then run the tools from there?

EDIT: Missed the `-Include` filter...
EDIT2: To top it off: I managed to mess up the bitmask in my draft PR. You will have to fetch the latest changes for this to work properly. Sorry!

**Reply by apantina** [reply]
No worries! Copying over the DLLs worked :) I fetched the latest changes and managed to verify an assertion successfully, thank you so much for all the help!
The assert file with a raw challenge (instead of the hash):
```
PS C:\Users\ardi\libfido2\build\x64\dynamic\tools\Release> cat ..\..\..\..\..\..\assert.txt
a2xWalpBQUFBQUFEcThVeGpwZm5LNkhUX3VvS211TEZWWnB4Ti1PVEdOZmY1ZF84WGwtMkFSMmN5NWRWTTNoYTZtaWhveDVmSnJ2MkVSMVdWVnVSZEJHRVdUSVlxSjh0RTEzTg==
internalfb.com
SO3HEwBHH9vI7D7aEoKrAYNLJ99p9SMYHIeHL8EcrlXZem/mFPSjFSEVIsXTetNa+lvIkbXofgDScWjKvwAMSw==
```

Creating an assertion via `fido2-assert.exe`:
```
.\fido2-assert.exe -G -w -tpin=false -i ..\..\..\..\..\..\assert.txt windows://hello
7h/k54HhOa1NUdMfoe5J2xLmTFMtAEoAgD4JpsBv/hQ=
internalfb.com
WCX+8b41zaVs59RYeoohrljoqwHBidr6NkPD0RevZ1AHVgUAAARa
MEUCIBvje+5GAUzwZyYv9VaUD6UdafRfZlL0svRinqyiby7AAiEA814/DeSK63hrI2QEJyuEQYJv4lC0U7DWOfsv6CWdJ7M=
```

I have a followup question though - any idea why I get asked for my Yubikey pin? I explicitly set `-tpin=false` and I still get the prompt (screenshot below). I also checked the debug info and I see `pin=false` being output.

```
 .\fido2-assert.exe -G -w -d -tpin=false -i ..\..\..\..\..\..\assert.txt windows://hello
client data:
  6b 6c 56 6a 5a 41 41 41 41 41 41 44 71 38 55 78
  6a 70 66 6e 4b 36 48 54 5f 75 6f 4b 6d 75 4c 46
  56 5a 70 78 4e 2d 4f 54 47 4e 66 66 35 64 5f 38
  58 6c 2d 32 41 52 32 63 79 35 64 56 4d 33 68 61
  36 6d 69 68 6f 78 35 66 4a 72 76 32 45 52 31 57
  56 56 75 52 64 42 47 45 57 54 49 59 71 4a 38 74
  45 31 33 4e
relying party id: internalfb.com
credential id:
  48 ed c7 13 00 47 1f db c8 ec 3e da 12 82 ab 01
  83 4b 27 df 69 f5 23 18 1c 87 87 2f c1 1c ae 55
  d9 7a 6f e6 14 f4 a3 15 21 15 22 c5 d3 7a d3 5a
  fa 5b c8 91 b5 e8 7e 00 d2 71 68 ca bf 00 0c 4b
up=omit
uv=omit
pin=false
webauthn_load: api version 4
cbor_decode_assert_authdata: buf=000001BD9A647770, len=37
7h/k54HhOa1NUdMfoe5J2xLmTFMtAEoAgD4JpsBv/hQ=
internalfb.com
WCX+8b41zaVs59RYeoohrljoqwHBidr6NkPD0RevZ1AHVgUAAARe
MEYCIQDNpz8xG1gNJmhL4Ro0yK34WzlyI5lsoUztwAte8/HXSwIhAMwuYpYcmnRcgX7uw0gmRr+DGmb/TPS8TdiQZPndRwUW
```

The prompt:

<img width="354" alt="image" src="https://github.com/Yubico/libfido2/assets/110903292/bed57886-e0e5-4d96-9149-22ada7d33f3c">

After putting in my pin I get the user presence check, pressing the yubikey closes the prompt and `fido2-assert` gives me back the signed assertion.

<img width="334" alt="image" src="https://github.com/Yubico/libfido2/assets/110903292/f50914da-794f-45a7-ac7d-fe6177776277">

**Reply by LDVG** [reply]
> I have a followup question though - any idea why I get asked for my Yubikey pin? I explicitly set -tpin=false and I still get the prompt (screenshot below). I also checked the debug info and I see pin=false being output.

Since it's not possible for a client application to pass a PIN to the `windows://hello` backend (it has its own prompt), we consider it to only support "built-in" user verification methods in an attempt to map it to libfido2 concepts. Try using the `-tuv=false` option instead; that should discourage the backend from prompting you for a PIN.

**Reply by apantina** [reply]
Thanks for the clarification. Adding `tuv=false` removed the PIN prompt!

**Reply by LDVG** [reply]
Thank you for testing, I'll try to finish up that PR sometime soon.
